File: 2022/04/solution.rs
use std::fs;

fn main() {
    let input = parse_input();
    println!("{}", part1(&input));
    println!("{}", part2(&input));
}

fn parse_input() -> Vec<[u32; 4]> {
    fs::read_to_string("input.txt")
        .unwrap()
        .lines()
        .map(|line| {
            let section_ids: Vec<_> = line
                .split(',')
                .flat_map(|range| {
                    range
                        .split('-')
                        .map(|section| section.parse::<u32>().unwrap())
                })
                .collect();
            [
                section_ids[0],
                section_ids[1],
                section_ids[2],
                section_ids[3],
            ]
        })
        .collect()
}

fn part1(input: &[[u32; 4]]) -> u32 {
    input
        .iter()
        .filter(|section_ids| {
            (section_ids[0] <= section_ids[2] && section_ids[1] >= section_ids[3])
                || (section_ids[0] >= section_ids[2] && section_ids[1] <= section_ids[3])
        })
        .count() as u32
}

fn part2(input: &[[u32; 4]]) -> u32 {
    input
        .iter()
        .filter(|section_ids| {
            (section_ids[0] >= section_ids[2] || section_ids[1] >= section_ids[2])
                && (section_ids[0] <= section_ids[3] || section_ids[1] <= section_ids[3])
        })
        .count() as u32
}

File: 2022/04/solution.py
def parse_input():
    with open('input.txt') as f:
        lines = f.readlines()
    return [[*map(int, line.replace(',', '-').split('-'))] for line in lines]


def part1(input):
    return sum(
        1 if (
            section_ids[0] <= section_ids[2] and section_ids[1] >= section_ids[3]
        ) or (
            section_ids[0] >= section_ids[2] and section_ids[1] <= section_ids[3]
        ) else 0
        for section_ids
        in input
    )


def part2(input):
    return sum(
        1 if (
            section_ids[0] >= section_ids[2] or section_ids[1] >= section_ids[2]
        ) and (
            section_ids[0] <= section_ids[3] or section_ids[1] <= section_ids[3]
        ) else 0
        for section_ids
        in input
    )


input = parse_input()
print(part1(input))
print(part2(input))
